V6.5.3 Redis[Clustered] - "CROSSSLOT Keys in request" when unlock

Hello SB-integration !

**In what version(s) of Spring Integration are you seeing this issue?**

Since 6.5.3.RELEASE (SpringBoot BOM parent : >=3.5.7)
6.5.2 tested and ok (SpringBoot BOM parent : <3.5.7)

**Describe the bug**

With a clustered redis when call the method Lock.unlock() 
Using implementation RedisLockRegistry.obtain("myKey") to obtain the lock

**To Reproduce**

Start a clustred Redis. 
Obtain a lock by calling RedisLockRegistry.obtain("myKey")
Call unlock method on obtained previous Lock

**Expected behavior**

The lock is created and visible with redis client (Like Redis insight).
When call unlock method. 
The name of the lock is sent on the dedicated redis channel.
The lock is removed

**Sample**


For now I can propose you this sample 

```kotlin

class Example(private val redisLockRegistry: RedisLockRegistry) {
   fun handle(): Void {
        val lock = redisLockRegistry.obtain("my-key-for-lock")
        val hasLock = lock.tryLock(35, TimeUnit.SECONDS)
        if (!hasLock) {
            throw NullPointerException("Unable to obtain lock")
        }
        val result = try {
          lock.unlock()
        }
    }
}

```

[Start with a project with Springboot 3.5.7 (minimum) ](https://start.spring.io/#!type=gradle-project&language=kotlin&platformVersion=3.5.7&packaging=jar&configurationFileFormat=properties&jvmVersion=17&groupId=com.example&artifactId=demo&name=demo&description=Demo%20project%20for%20Spring%20Boot&packageName=com.example.demo&dependencies=integration,data-redis-reactive)


I'm working to found a solution to push a repository with a sample. But get in mind the issue can be reproduced only with a redis cluster. 

Stacktrace : 

```text
Caused by: io.lettuce.core.RedisCommandExecutionException: CROSSSLOT Keys in request don't hash to the same slot
at io.lettuce.core.internal.ExceptionFactory.createExecutionException(ExceptionFactory.java:151)
at io.lettuce.core.internal.ExceptionFactory.createExecutionException(ExceptionFactory.java:120)
at io.lettuce.core.protocol.AsyncCommand.completeResult(AsyncCommand.java:124)
at io.lettuce.core.protocol.AsyncCommand.complete(AsyncCommand.java:115)
at io.lettuce.core.protocol.CommandWrapper.complete(CommandWrapper.java:67)
at io.lettuce.core.protocol.CommandWrapper.complete(CommandWrapper.java:67)
at io.lettuce.core.cluster.ClusterCommand.complete(ClusterCommand.java:50)
at io.lettuce.core.protocol.CommandWrapper.complete(CommandWrapper.java:67)
at io.lettuce.core.protocol.CommandHandler.complete(CommandHandler.java:769)
at io.lettuce.core.protocol.CommandHandler.decode(CommandHandler.java:704)
at io.lettuce.core.protocol.CommandHandler.channelRead(CommandHandler.java:621)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)
at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:501)
at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:399)
at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
... 1 common frames omitted\n"
```


A first investigation on our own : 

The stack trace come from here : https://github.com/spring-projects/spring-integration/blob/e73ab48769f2d996992228c106d5dbb97d1cbfd8/spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java#L713-L715


in 6.5.2 
The channel name is pass throught ARGV (arguments array) in 6.5.3 [the channel name for unlock is concat with the lock key name and sent as a key](https://github.com/spring-projects/spring-integration/blob/e73ab48769f2d996992228c106d5dbb97d1cbfd8/spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java#L712) .

--- a/spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
+++ b/spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
@@ -670,7 +670,7 @@ public final class RedisLockRegistry
 		private static final String UNLINK_UNLOCK_SCRIPT = """
 				local lockClientId = redis.call('GET', KEYS[1])
 				if (lockClientId == ARGV[1] and redis.call('UNLINK', KEYS[1]) == 1) then
-					redis.call('PUBLISH', KEYS[2], KEYS[1])
+					redis.call('PUBLISH', ARGV[2], KEYS[1])
 					return true
 				end
 				return false
@@ -693,8 +693,8 @@ public final class RedisLockRegistry
 		protected boolean removeLockKeyInnerUnlink() {
 			final String unLockChannelKey = RedisLockRegistry.this.unLockChannelKey + ":" + this.lockKey;
 			return Boolean.TRUE.equals(RedisLockRegistry.this.redisTemplate.execute(
-					UNLINK_UNLOCK_REDIS_SCRIPT, List.of(this.lockKey, unLockChannelKey),
-					RedisLockRegistry.this.clientId));
+					UNLINK_UNLOCK_REDIS_SCRIPT, List.of(this.lockKey),
+					RedisLockRegistry.this.clientId, unLockChannelKey));
 		}
 
 		private boolean subscribeLock(long time, long expireAfter) throws ExecutionException, InterruptedException {

--- a/src/reference/antora/modules/ROOT/pages/redis.adoc
+++ b/src/reference/antora/modules/ROOT/pages/redis.adoc
@@ -866,3 +866,25 @@ When it is set, the lock will be automatically renewed every `1/3` of the expira
 Starting with version 7.0, the `RedisLock` implements `DistributedLock` interface to support the feature of customized time-to-live (TTL) for the lock status data.
 A `RedisLock` can now be acquired using the `lock(Duration ttl)` or `tryLock(long time, TimeUnit unit, Duration ttl)` method, with a specified time-to-live (TTL) value.
 The `RedisLockRegistry` now provides new `renewLock(Object lockKey, Duration ttl)` method, allowing you to renew the lock with a custom time-to-live value.
+
+[[elasticache-valkey-cluster]]
+=== AWS ElastiCache for Valkey Support in cluster mode
+
+Starting with version 6.4.9/6.5.4/7.0.0, `RedisLockRegistry` supports AWS Elasticache for Valkey in cluster mode. In this version of valkey (a redis drop-in replacement), all PubSub operations (`PUBLISH`, `SUBSCRIBE`, etc.) use their sharded variants (`SPUBLISH`, `SSUBSCRIBE`, etc.) internally.
+If you are observing errors in the form of
+
+[source]
+----
+Caused by: io.lettuce.core.RedisCommandExecutionException: ERR Script attempted to access keys that do not hash to the same slot script: b2dedc0ab01c17f9f20e3e6ddb62dcb6afbed0bd, on @user_script:3.
+----
+
+in the `unlock` step of the `RedisLockRegistry`,  you have to supply a lock key that includes a hash tag `{...}` to ensure all operations in the `unlock` script are hashed to the same cluster slot/shard, e.g.:
+
+[source]
+----
+RedisLockRegistry lockRegistry = new RedisLockRegistry("my-lock-key{choose_your_tag}");
+
+lockRegistry.lock();
+# critical section
+lockRegistry.unlock();
+----

--- a/src/reference/antora/modules/ROOT/pages/redis.adoc
+++ b/src/reference/antora/modules/ROOT/pages/redis.adoc
@@ -870,8 +870,9 @@ The `RedisLockRegistry` now provides new `renewLock(Object lockKey, Duration ttl
 [[elasticache-valkey-cluster]]
 === AWS ElastiCache for Valkey Support in cluster mode
 
-Starting with version 6.4.9/6.5.4/7.0.0, `RedisLockRegistry` supports AWS Elasticache for Valkey in cluster mode. In this version of valkey (a redis drop-in replacement), all PubSub operations (`PUBLISH`, `SUBSCRIBE`, etc.) use their sharded variants (`SPUBLISH`, `SSUBSCRIBE`, etc.) internally.
-If you are observing errors in the form of
+Starting with version 6.4.9/6.5.4/7.0.0, `RedisLockRegistry` supports AWS Elasticache for Valkey in cluster mode.
+In this version of valkey (a redis drop-in replacement), all PubSub operations (`PUBLISH`, `SUBSCRIBE`, etc.) use their sharded variants (`SPUBLISH`, `SSUBSCRIBE`, etc.) internally.
+If you are observing errors in the form of:
 
 [source]
 ----

